src/WebJobs.Script/Host/FunctionMetadataManager.cs

```diff
@@ -214,12 +218,36 @@ private void LoadCustomProviderFunctions(List<FunctionMetadata> functionMetadata
 
         private void AddMetadataFromCustomProviders(IEnumerable<IFunctionProvider> functionProviders, List<FunctionMetadata> functionMetadataList)
         {
-            _logger.ReadingFunctionMetadataFromProvider(_metadataProviderName);
+            _logger.ReadingFunctionMetadataFromProvider(MetadataProviderName);
 
             var functionProviderTasks = new List<Task<ImmutableArray<FunctionMetadata>>>();
+
             foreach (var functionProvider in functionProviders)
             {
-                functionProviderTasks.Add(functionProvider.GetFunctionMetadataAsync());
+                var getFunctionMetadataFromProviderTask = functionProvider.GetFunctionMetadataAsync();
+                var delayTask = Task.Delay(TimeSpan.FromSeconds(MetadataProviderTimeoutInSeconds));
+
+                var completedTask = Task.WhenAny(getFunctionMetadataFromProviderTask, delayTask).ContinueWith(t =>
+                {
+                    if (t.Result == getFunctionMetadataFromProviderTask)
+                    {
+                        if (getFunctionMetadataFromProviderTask.IsFaulted)
+                        {
+                            // Task completed but with an error
+                            _logger.LogWarning($"Failure in retrieving metadata from '{functionProvider.GetType().FullName}': {getFunctionMetadataFromProviderTask.Exception?.Flatten().ToString()}");
+                            return ImmutableArray<FunctionMetadata>.Empty;
```

Comment:
Was the behavior difference considered here? Prior to this change, this would throw an exception that would bubble up in the metadata loading path (failing initialization). This will now log and move on, initializing with an empty list of functions.

Reply:
Removed this behavior change. In the latest iteration, the timeout case will throw a `TimeOutException`